core/controllers/reader.py

```diff
@@ -72,97 +70,29 @@ def test_can_play(self, exploration_id, **kwargs):
-                'answer_group_index': best_matched_answer_group_index,
-                'classification_certainty': best_matched_truth_value,
-                'rule_spec_index': best_matched_rule_spec_index,
-            }
-
-    return None
-
-
-def classify_soft_rule(state, params, input_type, normalized_answer, fs):
-    """Find the maximum soft rule that matches. This is done by ORing
-    (maximizing) all truth values of all rules over all answer groups. The
-    group with the highest truth value is considered the best match.
-    """
-    best_matched_answer_group = None
-    best_matched_answer_group_index = len(state.interaction.answer_groups)
-    best_matched_rule_spec_index = None
-    best_matched_truth_value = rule_domain.CERTAIN_FALSE_VALUE
-
-    for (answer_group_index, answer_group) in enumerate(
-            state.interaction.answer_groups):
-        fuzzy_rule_spec_index = answer_group.get_fuzzy_rule_index()
-        if fuzzy_rule_spec_index is not None:
-            fuzzy_rule_spec = answer_group.rule_specs[fuzzy_rule_spec_index]
-        else:
-            fuzzy_rule_spec = None
-        if fuzzy_rule_spec is not None:
-            evaluated_truth_value = rule_domain.evaluate_rule(
-                fuzzy_rule_spec, input_type, params, normalized_answer, fs)
-            if evaluated_truth_value == rule_domain.CERTAIN_TRUE_VALUE:
-                best_matched_truth_value = evaluated_truth_value
-                best_matched_rule_spec_index = fuzzy_rule_spec_index
-                best_matched_answer_group = answer_group
-                best_matched_answer_group_index = answer_group_index
-                return {
-                    'outcome': best_matched_answer_group.outcome.to_dict(),
-                    'answer_group_index': best_matched_answer_group_index,
-                    'classification_certainty': best_matched_truth_value,
-                    'rule_spec_index': best_matched_rule_spec_index,
-                }
-
-    return None
-
-
 def classify_string_classifier_rule(state, normalized_answer):
     """Run the classifier if no prediction has been made yet. Currently this
     is behind a development flag.
     """
     best_matched_answer_group = None
     best_matched_answer_group_index = len(state.interaction.answer_groups)
     best_matched_rule_spec_index = None
-    best_matched_truth_value = rule_domain.CERTAIN_FALSE_VALUE
+    best_matched_truth_value = feconf.CERTAIN_FALSE_VALUE
 
     sc = classifier_services.StringClassifier()
     training_examples = [
         [doc, []] for doc in state.interaction.confirmed_unclassified_answers]
     for (answer_group_index, answer_group) in enumerate(
             state.interaction.answer_groups):
-        fuzzy_rule_spec_index = answer_group.get_fuzzy_rule_index()
-        if fuzzy_rule_spec_index is not None:
-            fuzzy_rule_spec = answer_group.rule_specs[fuzzy_rule_spec_index]
+        classifier_spec_index = answer_group.get_classifier_index()
```

Comment:
I'm not so sure about this naming. Fuzzy rules are not the same thing as classifiers. They are still technically rules which contain training data used to build statistical classifiers. This might need more thought that goes outside the scope of this PR. I'm not currently recalling what's supposed to happen with fuzzy rules.